File: lib/src/models/person.dart
import 'package:cloud_firestore/cloud_firestore.dart';

import 'enums.dart';

/// Pessoa (atleta, técnico, árbitro, organizador).
///
/// Entidade central que pode ter múltiplos papéis.
class Person {
  final String id;
  final String name;
  final String? email;
  final String? phone;
  final String? photoUrl;
  final Gender? gender;
  final DateTime? birthDate;
  final String? computedAgeGroup;
  final List<PersonRole> roles;
  final String status;
  final DateTime createdAt;
  final DateTime? updatedAt;

  const Person({
    required this.id,
    required this.name,
    this.email,
    this.phone,
    this.photoUrl,
    this.gender,
    this.birthDate,
    this.computedAgeGroup,
    this.roles = const [],
    this.status = 'ACTIVE',
    required this.createdAt,
    this.updatedAt,
  });

  /// Cria Person a partir de documento Firestore.
  factory Person.fromFirestore(DocumentSnapshot doc) {
    final data = doc.data() as Map<String, dynamic>;
    return Person(
      id: doc.id,
      name: data['name'] as String? ?? '',
      email: data['email'] as String?,
      phone: data['phone'] as String?,
      photoUrl: data['photoUrl'] as String?,
      gender: genderFromString(data['gender'] as String?),
      birthDate: data['birthDate'] != null
          ? (data['birthDate'] as Timestamp).toDate()
          : null,
      computedAgeGroup: data['computedAgeGroup'] as String?,
      roles: (data['roles'] as List<dynamic>?)
              ?.map((r) => PersonRole.values.firstWhere(
                    (e) => e.name == r,
                    orElse: () => PersonRole.athlete,
                  ))
              .toList() ??
          [],
      status: data['status'] as String? ?? 'ACTIVE',
      createdAt: (data['createdAt'] as Timestamp?)?.toDate() ?? DateTime.now(),
      updatedAt: (data['updatedAt'] as Timestamp?)?.toDate(),
    );
  }

  /// Converte para Map (Firestore).
  Map<String, dynamic> toFirestore() {
    return {
      'name': name,
      if (email != null) 'email': email,
      if (phone != null) 'phone': phone,
      if (photoUrl != null) 'photoUrl': photoUrl,
      if (gender != null) 'gender': gender!.name,
      if (birthDate != null) 'birthDate': Timestamp.fromDate(birthDate!),
      if (computedAgeGroup != null) 'computedAgeGroup': computedAgeGroup,
      'roles': roles.map((r) => r.name).toList(),
      'status': status,
      'createdAt': FieldValue.serverTimestamp(),
      'updatedAt': FieldValue.serverTimestamp(),
    };
  }

  /// Calcula ageGroup a partir de birthDate.
  String? calculateAgeGroup() {
    if (birthDate == null) return null;
    final now = DateTime.now();
    final age = now.year - birthDate!.year -
        (now.month < birthDate!.month ||
                (now.month == birthDate!.month && now.day < birthDate!.day)
            ? 1
            : 0);
    if (age <= 6) return 'U6';
    if (age <= 8) return 'U8';
    if (age <= 10) return 'U10';
    if (age <= 12) return 'U12';
    if (age <= 14) return 'U14';
    if (age <= 16) return 'U16';
    if (age <= 18) return 'U18';
    if (age <= 20) return 'U20';
    return 'OPEN';
  }

  Person copyWith({
    String? name,
    String? email,
    String? phone,
    String? photoUrl,
    Gender? gender,
    DateTime? birthDate,
    String? computedAgeGroup,
    List<PersonRole>? roles,
    String? status,
  }) {
    return Person(
      id: id,
      name: name ?? this.name,
      email: email ?? this.email,
      phone: phone ?? this.phone,
      photoUrl: photoUrl ?? this.photoUrl,
      gender: gender ?? this.gender,
      birthDate: birthDate ?? this.birthDate,
      computedAgeGroup: computedAgeGroup ?? this.computedAgeGroup,
      roles: roles ?? this.roles,
      status: status ?? this.status,
      createdAt: createdAt,
      updatedAt: updatedAt,
    );
  }
}

File: src/main/java/br/com/flagplatform/common/model/Person.java
package br.com.flagplatform.common.model;

import com.google.cloud.firestore.annotation.DocumentId;
import com.google.cloud.firestore.annotation.ServerTimestamp;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.Instant;
import java.time.LocalDate;
import java.util.List;

/**
 * Pessoa (atleta, técnico, árbitro, organizador).
 * 
 * Entidade central que pode ter múltiplos papéis.
 * Mapeado para a collection 'persons' no Firestore.
 */
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Person {

    @DocumentId
    private String id;

    private String name;

    private String email;

    private String phone;

    private String photoUrl;

    private String gender; // MALE, FEMALE, MIXED

    private LocalDate birthDate;

    private String computedAgeGroup; // U6, U8, U10, U12, U14, U16, U18, U20, OPEN

    @Builder.Default
    private List<String> roles = List.of(); // ATHLETE, COACH, MANAGER, ORGANIZER, REFEREE, ADMIN

    @Builder.Default
    private String status = "ACTIVE"; // ACTIVE, INACTIVE

    @ServerTimestamp
    private Instant createdAt;

    @ServerTimestamp
    private Instant updatedAt;

    /**
     * Calcula ageGroup a partir de birthDate.
     */
    public String calculateAgeGroup() {
        if (birthDate == null) {
            return null;
        }
        
        LocalDate now = LocalDate.now();
        int age = now.getYear() - birthDate.getYear();
        
        if (now.getMonthValue() < birthDate.getMonthValue() ||
            (now.getMonthValue() == birthDate.getMonthValue() && now.getDayOfMonth() < birthDate.getDayOfMonth())) {
            age--;
        }

        if (age <= 6) return "U6";
        if (age <= 8) return "U8";
        if (age <= 10) return "U10";
        if (age <= 12) return "U12";
        if (age <= 14) return "U14";
        if (age <= 16) return "U16";
        if (age <= 18) return "U18";
        if (age <= 20) return "U20";
        return "OPEN";
    }
}
